functorch/functorch/csrc/BatchRulesHelper.cpp

```diff
@@ -147,4 +147,53 @@ void run_jit_decomposition(const c10::OperatorHandle& op, torch::jit::Stack* sta
   }
 }
 
+static void handleScalarTypePromotion(Tensor& logical_scalar_tensor, Tensor& second) {
+  auto result_type = at::native::result_type(logical_scalar_tensor[0], second);
+  if (logical_scalar_tensor.scalar_type() != result_type) {
+    logical_scalar_tensor = logical_scalar_tensor.to(result_type);
+  }
+  if (second.scalar_type() != result_type) {
+    second = second.to(result_type);
+  }
+}
+
+std::tuple<Tensor, Tensor> _binary_pointwise_helper(
+    const Tensor& tensor, optional<int64_t> tensor_batch_dim,
+    const Tensor& other, optional<int64_t> other_batch_dim,
+    bool do_type_promotion) {
+  // compute max logical rank
+  auto tensor_logical_rank = rankWithoutBatchDim(tensor, tensor_batch_dim);
+  auto other_logical_rank = rankWithoutBatchDim(other, other_batch_dim);
+  auto max_logical_rank = std::max(tensor_logical_rank, other_logical_rank);
+
+  auto tensor_ = moveBatchDimToFront(tensor, tensor_batch_dim);
+  auto other_ = moveBatchDimToFront(other, other_batch_dim);
+
+  // In the (0D, ND) case, type promotion semantics are different :/
+  if (do_type_promotion) {
+    auto tensor_is_logical_scalar = (tensor_logical_rank == 0 && tensor_batch_dim.has_value());
+    auto other_is_logical_scalar = (other_logical_rank == 0 && other_batch_dim.has_value());
+    if (tensor_is_logical_scalar && !other_is_logical_scalar) {
+      handleScalarTypePromotion(tensor_, other_);
+    }
+    if (other_is_logical_scalar && !tensor_is_logical_scalar) {
+      handleScalarTypePromotion(other_, tensor_);
+    }
+  }
+
+  // If the dimensions aren't aligned, we need to line them up.
+  // Tensor[B, 3] + Tensor[2, 5, 3] -> Tensor[B, 1, 1, 3] + Tensor[2, 5, 3]
+  // Note that only tensors that have a batch dim need to be modified.
+  // Tensor[B, 2, 3, 5] + Tensor[5] -> no changes needed
+  tensor_ = maybePadToLogicalRank(tensor_, tensor_batch_dim, max_logical_rank);
+  other_ = maybePadToLogicalRank(other_, other_batch_dim, max_logical_rank);
+
+  return std::make_tuple(tensor_, other_);
+}
+
+std::tuple<Tensor, Tensor> _binary_pointwise_helper(
+    const Tensor& tensor, optional<int64_t> tensor_batch_dim,
+    const Tensor& other, optional<int64_t> other_batch_dim) {
+  return _binary_pointwise_helper(tensor, tensor_batch_dim, other, other_batch_dim, true);
```

Comment:
nit: leave a comment for boolean arguments to make it easier to read, e.g. `/*do_type_promotion*/ true)`